File: sast-python2/case/accuracy/field_sensitive/class/field_len_001_T.py
# -*- coding: utf-8 -*-
# evaluation information start
# real case = true
# evaluation item = 准确度->对象敏感与域敏感分析->区分不同类对象的不同字段
# scene introduction = 路径长度
# level = 3
# bind_url = accuracy/field_sensitive/class/field_len_001_T
# evaluation information end
import os

def field_len_001_T(taint_src):
    class A(object):
        def __init__(self, taint_src):
            self.b = B(taint_src)  # 将参数传递给下一层类

    class B(object):
        def __init__(self, taint_src):
            self.c = C(taint_src)  # 继续传递参数到 C

    class C(object):
        def __init__(self, taint_src):
            self.data = taint_src   # 使用参数赋值给 data
            self.sani = u'_'

    # 创建实例时传入 taint_src
    a = A(taint_src)
    taint_sink(a.b.c.data)        # 通过路径访问 data 属性


def taint_sink(o):
    os.system(o)


# 示例调用
if __name__ == u"__main__":
    taint_src = u"taint_src_value"
    field_len_001_T(taint_src)

File: sast-php/case/accuracy/object_field_sensitive/field_len/field_len_001_T.php
<?php
// evaluation information start
// real case = true
// evaluation item = 准确度->对象域敏感->域长度
// scene introduction = 三层嵌套对象字段访问，末端持有污染数据
// level = 2
// bind_url = accuracy/object_field_sensitive/field_len/field_len_001_T
// evaluation information end

function __taint_sink($o) {
    shell_exec($o);
}

class C001 {
    public $data;
    public $safe;

    public function __construct($v) {
        $this->data = $v;
        $this->safe = "safe";
    }
}

class B001 {
    public $c;

    public function __construct($v) {
        $this->c = new C001($v);
    }
}

class A001 {
    public $b;

    public function __construct($v) {
        $this->b = new B001($v);
    }
}

function field_len_001_T($__taint_src) {
    $a = new A001($__taint_src);
    __taint_sink($a->b->c->data);
}

$taint_src = "taint_src_value";
field_len_001_T($taint_src);
